### DISCUSSION TITLE
Using `None` in form field declarations with Pydantic

### DISCUSSION THREAD

**Discussion by 9en9i** [fastapi/fastapi#12150] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Form
from pydantic import BaseModel

app = FastAPI()

class NewForm(BaseModel):
    field: int | None

@app.post("/api")
def func(form: NewForm = Form()):
    return form.model_dump(exclude_unset=True)
```

### Description

I have encountered an inability to pass a None value to the form. I tried sending “null” or not filling in the value, but none of these things worked. Either way I get the error `Input should be a valid integer, unable to parse string as an integer`.

<img width="1127" alt="image" src="https://github.com/user-attachments/assets/0d8bb67f-b549-441c-9821-715974313e6d">

### Operating System

macOS

### FastAPI Version

0.113.0

### Pydantic Version

2.9.0

### Python Version

Python 3.12.2

**Comment by TomFaulkner** [comment]
I ran into the same problem with Firefox sends an empty optional form field as an empty string. The Pydantic model has the field as `datetime.date | None`. I agree with Pydantic that this isn't either a date or None, but unfortunately, browsers don't deal with types well in forms.

I think the solution is probably to add parsing for any optional types to parse an empty string as None. Unfortunately, that seems worse than the pre 0.113.0 alternative.

**Reply by 9en9i** [reply]
I have in my project (even before 0.113.0 was released) we are using Pydantic models for forms (using a workaround). And as far as I know, from frontend we get only js-value null and cast null string to None. Maybe should use something similar to support None in models

**Reply by TomFaulkner** [reply]
Is that from a browser without any JS modifying it? I just checked Firefox and Chromium, both send empty strings. Or, rather, with the raw data it's `foo=bar&date_field=&baz=foo2`.

Prior to 0.113.0 I have been using this: https://gist.github.com/TomFaulkner/eeb8028d4afaca35dda85ebe5e7b76ef
(I don't use `form_body` as a decorator to preserve access to the model, but it works well enough.)

I'm using htmx for the frontend, I'm sure I could write something that would prevent sending empty strings or change them to null, but that is definitely less than ideal.

**Reply by 9en9i** [reply]
On the frontend js framework, we create a `FormData` object in the code, write null (not a string, js literal) to it, and we get a string “null” on the backend which we convert to `None`. I agree that this is not a perfect solution, but I don't have a better one. It turns out that from the frontend developer side everything looks good, but on the backend we have to use workaround.

**Reply by pythonweb2** [reply]
Why not add a pydantic `BeforeValidator` that will convert an empty string to `None`?

**Reply by 9en9i** [reply]
Yes, I just came to the point of adding such code to the base class:

```python
    @model_validator(mode="before")
    @staticmethod
    def convert_null(obj: dict[str, object]) -> dict[str, object]:
        return {k: None if v == "null" else v for k, v in obj.items()}
```

But it seems like it would be nice to have `None` support in Pydantic forms in the framework, though I may be wrong.

UPD: this does not work with the `UploadFile` field

**Reply by TomFaulkner** [reply]
> Why not add a pydantic `BeforeValidator` that will convert an empty string to `None`?

That's definitely doable. I'll probably do something along those lines for new models and convert the others later.

**Reply by Nagidrop** [reply]
> Yes, I just came to the point of adding such code to the base class:
> 
> ```python
>     @model_validator(mode="before")
>     @staticmethod
>     def convert_null(obj: dict[str, object]) -> dict[str, object]:
>         return {k: None if v == "null" else v for k, v in obj.items()}
> ```
> 
> But it seems like it would be nice to have `None` support in Pydantic forms in the framework, though I may be wrong.
> 
> UPD: this does not work with the `UploadFile` field

Same workaround but cleaner, 1 field only
```python
@field_validator("field_name", mode="before")
def any_function_name(cls, value: str) -> str | None:
    return value if value else None  # None for empty strings, or any other conditions
```
